docs/references/architecture/adr-119-dog-mempool-gossip.md

```diff
@@ -0,0 +1,239 @@
+
+## Alternative Approaches
+
+> This section contains information around alternative options that are considered
+> before making a decision. It should contain a explanation on why the alternative
+> approach(es) were not chosen.
+
+## Decision
+
+The CometBFT team has decided to implement the protocol on top of the existing mempool
+(`flood` ) dissemination protocol but make it optional. The protocol can be activated/deactivate
+via a new config flag `enable_dog_protocol`. More details on this in the sections below. 
+
+## Detailed Design
+
+We start the design section with a general description of how the algorithm works. 
+
+CometBFT, at the moment, caches received transactions. For all transactions, CometBFT stores the IDs of all the peers that have sent it in a `senders` list. When a transaction enters the mempool,
+if it was already received it can be retrieved from the cache. When this happens a certain number of times, the receiving node notifies the sender to stop forwarding any transaction from the peer 
+that had originally sent the transaction. The exact point when the notification is sent is defined below as a redundancy threshold. (TODO LiNK)
+
+Namely, let's assume that transaction `tx_1` was sent by node `B` to node `C` which forwarded it to `A`. `A` already has the transaction and notifies node `C` about this. Node `C` will look up the senders of transaction `tx_1` and in the future disable transaction forwarding from node `B` to node `A`.
+
+In reality, if node `C` ha received `tx_1` from multiple senders (`X`,`Y`, `B`), it will pick the first one that has sent it and disable that route. 
+
+When a connection to a peer is lost a node sends a `Reset` message to its peers. It
+signals to the peers that they should delete all routing information related to this node. This assures that, in case the network topology changes, or a peer becomes unavailable, the node is able to discover transactions via a previously disabled route. 
+
+### Redundancy control
+
+In an ideal setting, receiving a transaction more than once is not needed. But in a Byzantine setting, allowing for only one route for a transaction can introduce an attack vector. 
+
+Operators can thus configure a desired redundancy and the gossip protocol will adjust route disabling based on this setting. Redundancy is defined as the ratio between duplicate and unique transactions. Operators can also define how much they tolerate the redundancy to deviate from the desired level.
+
+By default, the redundancy is set to 1 and impacts the gossiping of transactions as follows:
+
+
+```
+  redundancy := duplicateTxs / firstTimeTxs
+  if redundancy < r.redundancyLowerBound:
+    peer.send(Reset)
+  if redundancy > redundancyUpperBound
+    enableRouteDisabling() 
+```
+
+Redundancy is adjusted whenever a new transaction is added into the mempool.
+(TODO - revisit whether this should also be done on duplicates.)
+
+
+### Impacted areas of code
```

Comment:
The `Mempool` interface will have a new method `GetSenders(txKey types.TxKey) ([]nodekey.ID, error)`, and the `Entry` interface, `Senders() []nodekey.ID`.